### ISSUE TITLE
Idea: support broadcasting transactions over tor

### ISSUE THREAD

**Issue by MoneroArbo** [cculianu/Fulcrum#139]
Rationale: Bitcoin Core can broadcast transactions over tor, but a full node connected via both i2p and clearnet broadcasts transactions over both so doesn't protect well against IP correlation. Currently few wallets support tor broadcasting within the wallet. Except for Electrum Personal Server, electrum servers don't support tor broadcasting, and EPS as the name suggests is not fit for a public electrum server.

Thus, an modern Electrum server fit for public use that supports broadcasting transactions over tor could increase transaction privacy for many users.

**Comment by cculianu** [comment]
I am not sure I understand. Fulcrum doesn't communicate with the bitcoin peer-to-peer network. It just talks to bitcoind. It only "broadcasts" via the `sendrawtransaction` RPC call to bitcoind. If bitcoind is connected to other nodes and they are tor, then it will broadcast over tor.

**Comment by cculianu** [comment]
Oh.. I see. You want Fulcrum to communicate via the p2p network itself and broadcast txns to the p2p network? Sorry, that's out of scope for this project.  I'm closing this issue but feel free to communicate with me further here if you like.

**Comment by cculianu** [comment]
>  RPC transactions only over tor but that doesn't seem likely to be implemented.

Yes, this really is a job for bitcoind to get right... Do they even have an issue requesting this?

As for Fulcrum communicating via p2p. Maybe someday but the bitcoin p2p protocol is fairly complex and I never intended Fulcrum to "speak" this protocol.  That being said -- I will re-open this and mark it as a feature request type of issue.  I can see why some users might want this feature someday.

**Comment by cculianu** [comment]
> Anyway, my alternate suggestion would be the option to hand off transactions to onion electrum peers instead of the local node when onion electrum peers are enabled.

Ah. Maybe that would be a good hack. At least minimize the privacy leak... Interesting idea.

**Comment by MoneroArbo** [comment]
> Yes, this really is a job for bitcoind to get right... Do they even have an issue requesting this?

Yes there's an open issue since mid 2020 that never really got traction: https://github.com/bitcoin/bitcoin/issues/19042

Comment from Feb of this year reads:

> The reason why I haven't drafted an implementation yet is because the wallet can't keep track of non-mempool change. See https://github.com/bitcoin/bitcoin/issues/11887. And fixing that appears non-trivial at first glance.

Basically, you need to hand off the transaction without putting it in your own mempool, which for reasons I didn't quite grok or dig into is apparently non-trivial.

> Ah. Maybe that would be a good hack. At least minimize the privacy leak... Interesting idea. It would require Fulcrum to behave like a client (which it already knows how to do anyway). Interesting.

Glad you think so :)

**Comment by cculianu** [comment]
Yeah I can see why they want to avoid implementing it.  It's kind of a can of worms that needs to be though about.  The actual implementation would just be a bool flag in `CTxMemPoolEntry` to track if it came from RPC and/or wallet or not, and if it did, only send INV's to tor peers.

But then some corner cases come up which need to be though about.  What happens if you are given a CHILD txn of such txns -- one that spends the outputs of this txn.  Do you broadcast that?  Or is this flag "inherited"?  My gut tells me it's ok to broadcast unconfirmed children of such an unconfirmed txn.  After all if you got it "from the outside" it means it originated somewhere else and you aren't leaking privacy then.  This is a corner case.

Also there is this MEMPOOL p2p message.  I guess you would need to suppress those txns from that message too... although that message is hardly used, and it has some DoS controls built-in.. it can be used to fingerprint nodes...  

Yeah it's not impossible just a lot of deep thinking to think about..

**Comment by cculianu** [comment]
I am leaving this note here for myself or for anyone in the future interested in thinking about this issue:

---

Note that if Fulcrum were to implement some light/dumb p2p protocol to just forward transactions it sees from clients via Tor -- there is no guarantee that those transactions are sane and/or otherwise meeting consensus rules.  

This means that the Fulcrum "light p2p node" would quickly find itself potentially banned from all peers if it is spamming nonsense txns to its peers.

So to avoid that, it would need to do first "check" with its local bitcoind if the txn it got from a client is "Safe"..  RPC call `sendrawtransaction` with the "test only" flag set would do this.

**Comment by MoneroArbo** [comment]
question: how would a .onion node block tor peers? because tor itself doesn't let onion sites learn the source of incoming connects afaik, and there's no exit nodes to block when connecting to a .onion. 

checking transactions for sanity seems good regardless to avoid inadvertent spamming, but from a privacy perspective I'm wondering if bitcoind sends uniquely identifying information to tor peers, and if fulcrum would need to do the same. If so, maybe just generate a new "peer ID" or whatever for each connection to protect against transaction correlation.

**Comment by cculianu** [comment]
> question: how would a .onion node block tor peers?

Oh. Duh. That's a good point. It can't, really. :)  My bad. I will edit my message.  Still.. it's always a good idea not to antagonize peers and get kicked.. even if it doesn't really mean a real ban..

> checking transactions for sanity seems good regardless to avoid inadvertent spamming, 

Yeah maybe that's really the point of my little note above. True.

> I'm wondering if bitcoind sends uniquely identifying information to tor peers, 

Well.. aside from "subversion" user agent string.. late incarnations of bitcoind go to GREAT lengths to not send any fingerprinting .. as much as they can do so.  Even stuff like sharing a list of known peers with its peers is super slow now -- each node basically only gives out a static list of peers it knows about that is updated like once every 24 hours it's insane.  I think even the MEMPOOL p2p command is rate-limited to something super crazy slow... so Core over the years has tried to minimize fingerprinting..

Yes I guess if Fulcrum were to participate in the p2p network it should be sneaky and "pretend" it is just another bitcoind node. Is that what you are getting at?  Have it as unlikely as possible that Fulcrum could be "detected" as anything in particular.  Yes, good point. I will update my note..

**Comment by MoneroArbo** [comment]
That, and also some sort of rotating through tor peers. Not sure what would be the best way to implement but something like, periodically get a peer list from the local node, establish a connections to some reasonable subset, then when a transaction is received send it to one of the connected peers at random. Additionally at some interval X, connect to new peers and drop old ones.

This sort of scheme would ensure there are always available connections for broadcasting while minimizing the chance that transactions can be identified as from the same source, fulcrum or otherwise. The dropping and making new connections I think is an important addition because even if you disconnect and reconnect to the same node, it won't be able to tell those connections are from the same source.

My initial thought was to establish a new connection for every single transaction, which is what wallets like Wasabi and Sparrow do, but a public electrum server might be handling too many transactions to make that reasonable.

Anyway I clearly I don't have a deep understanding of the Bitcoin p2p protocol so some of my assumptions might be off but hopefully you can see what I'm getting at.
